File: lab6-rectypes/c/table.c
#include <stdio.h>
#include <malloc.h>

typedef enum {TABLE, TBODY, TR, TD } HTMLElement;

typedef struct
{ 
    HTMLElement element;
    int childrenNumber;
    struct HTMLFR * children;
      // incomplete struct, completion below
      // (struct HTMLFR = HTMLFragment)
}
HTMLFrElement; // the new record type

typedef struct
{
    char * text;
}
HTMLFrText;

typedef union
{
    HTMLFrElement frElement; // either this
    HTMLFrText ftText; // or this (a string)
}
HTMLFRUnion;

typedef enum {FRElem, FRText} HTMLFRRepr;

typedef struct HTMLFR // struct named for completion of above
{
    HTMLFRRepr repr; // discriminant
    HTMLFRUnion content; // the union
} 
HTMLFragment;

HTMLFragment * makeTable(char *** rows)
{
    HTMLFragment * table; // pointer type
    
    // allocate space for the record:
    table = 
        (HTMLFragment *) malloc(sizeof(HTMLFragment));
          // this is type casting of the pointer returned
          // by malloc
          
    // initialise the table element record:
    // (* table) is the memory cell pointed to by table
    (*table).repr = FRElem;
    (*table).content.frElement.element = TABLE;
    (*table).content.frElement.childrenNumber = 1;
    
    // create the child element - tbody
    HTMLFragment * tbody;
    tbody = 
        (HTMLFragment *) malloc(sizeof(HTMLFragment));
    // put tbody into the children list of table
    (*table).content.frElement.children = tbody;
    
    // initialise tbody record... etc.
    (*tbody).content.frElement.element = TBODY;
    (*tbody).content.frElement.childrenNumber = 0; // TODO
    
    // TODO: add all rows to tbody
    // need to count the rows first
    // then allocate space for children array
    // then create tr elements one by one and put pointers
    //  to them in the children array
    // each tr element will have its own td children...
    
    // working with tree structures in C is a nightmare
    
    return table;
}

int countNodes(HTMLFragment * fragment)
{
    int result = 1;
    
    if ( (*fragment).repr == FRElem )
    {
        HTMLFrElement element = (*fragment).content.frElement;
        int i;
        HTMLFragment * childPtr = element.children;
        for ( i = 0; i < element.childrenNumber; i++ )
        {
            result = result + countNodes(childPtr);
            childPtr = childPtr + 1; // pointer shift
        }
    }
    
    return result;
}

char rows1[2][4][4] =
     {{"X","X","O",0},{"O","B","X",0},{"O","B","B",0},0};

int main(char** argv, int argc)
{
    int n = countNodes(makeTable(&rows1));
    printf("countNodes = %d\n", n);
    return 0;
}

File: lab6-rectypes/java/Table.java
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;

public class Table
{

    private static abstract class HTMLFragment 
    {
        public int countNodes()
        {
            int result = 1;
            
            if ( this instanceof HTMLFrElement )
            {
                for ( HTMLFragment child : ((HTMLFrElement)this).children )
                {
                    result += child.countNodes();
                }
            }
            
            return result;
        }
    }
    
    private static enum HTMLElement 
    { TABLE, TBODY, TR, TD }
    
    private static class HTMLFrElement extends HTMLFragment
    {
        public HTMLElement element;
        public List<HTMLFragment>  children;
        
        public HTMLFrElement(HTMLElement element)
        {
            this.element = element;
            this.children = new LinkedList<HTMLFragment>();
        }
        
        public String toString()
        {
            String result = "";
            result += "<" + element + ">";
            for ( HTMLFragment child : children)
            {
                result += child;
            }
            result += "</" + element + ">";
            return result;
        }
    }
    
    private static class HTMLFrText extends HTMLFragment
    {
        public String text;
        
        public HTMLFrText(String text)
        {
            this.text = text;
        }
        
        public String toString()
        {
            return text;
        }
    }
    
    private static HTMLFragment makeTable(String[][] rows)
    {
        HTMLFrElement table = 
            new HTMLFrElement(HTMLElement.TABLE);
        
        HTMLFrElement tbody = 
            new HTMLFrElement(HTMLElement.TBODY);
        table.children.add(tbody);
        
        for (String[] row : rows)
        {
            HTMLFrElement tr =
                new HTMLFrElement(HTMLElement.TR);
            tbody.children.add(tr);
            
            for (String cellValue : row)
            {
                HTMLFrElement td =
                    new HTMLFrElement(HTMLElement.TD);
                tr.children.add(td);
                
                td.children.add(new HTMLFrText(cellValue));
            }
        }
        
        return table;
    }
    
    public static void main(String[] args)
    {
        HTMLFragment htmlBoard1 =
            makeTable(new String[][] {{"X","X"},
                                      {"O","B"}});
        
        System.out.println
            ("htmlBoard1.countNodes() = " + htmlBoard1.countNodes());
        
//        HTMLFrElement row =
//            new HTMLFrElement(HTMLElement.TR);
//        HTMLFrElement cell =
//            new HTMLFrElement(HTMLElement.TD);
//        cell.children.add(new HTMLFrText("B"));
//        row.children.add(cell);
//        row.children.add(cell);
//        row.children.add(cell);
//        System.out.println("row.countNodes() = " + row.countNodes());
    }
}